src/dbnode/storage/series/buffer.go

```diff
@@ -181,287 +161,281 @@ func (b *dbBuffer) Write(
 	value float64,
 	unit xtime.Unit,
 	annotation []byte,
+	wopts WriteOptions,
 ) error {
-	now := b.nowFn()
-	futureLimit := now.Add(1 * b.bufferFuture)
-	pastLimit := now.Add(-1 * b.bufferPast)
-	if !futureLimit.After(timestamp) {
-		return m3dberrors.ErrTooFuture
-	}
-	if !pastLimit.Before(timestamp) {
-		return m3dberrors.ErrTooPast
+	wType := wopts.WriteType
+	if wType == UndefinedWriteType {
+		wType = b.writeType(timestamp, b.nowFn())
 	}
 
-	bucketStart := timestamp.Truncate(b.blockSize)
-	idx := b.writableBucketIdx(timestamp)
-	if b.buckets[idx].needsReset(bucketStart) {
-		// Needs reset
-		b.DrainAndReset()
+	if wType == ColdWrite && !b.coldWritesEnabled {
+		return m3dberrors.ErrColdWritesNotEnabled
 	}
 
-	return b.buckets[idx].write(timestamp, value, unit, annotation)
+	blockStart := timestamp.Truncate(b.blockSize)
+	buckets := b.bucketsAtCreate(blockStart)
+	b.putBucketInCache(buckets)
+	return buckets.write(timestamp, value, unit, annotation, wType)
 }
 
-func (b *dbBuffer) writableBucketIdx(t time.Time) int {
-	return int(t.Truncate(b.blockSize).UnixNano() / int64(b.blockSize) % bucketsLen)
-}
+func (b *dbBuffer) writeType(timestamp time.Time, writeTime time.Time) WriteType {
+	ropts := b.opts.RetentionOptions()
+	futureLimit := writeTime.Add(1 * ropts.BufferFuture())
+	pastLimit := writeTime.Add(-1 * ropts.BufferPast())
 
-func (b *dbBuffer) IsEmpty() bool {
-	canReadAny := false
-	for i := range b.buckets {
-		canReadAny = canReadAny || b.buckets[i].canRead()
+	if pastLimit.Before(timestamp) && futureLimit.After(timestamp) {
+		return WarmWrite
 	}
-	return !canReadAny
-}
 
-func (b *dbBuffer) Stats() bufferStats {
-	var stats bufferStats
-	writableIdx := b.writableBucketIdx(b.nowFn())
-	for i := range b.buckets {
-		if !b.buckets[i].canRead() {
-			continue
-		}
-		if i == writableIdx {
-			stats.openBlocks++
-		}
-		stats.wiredBlocks++
-	}
-	return stats
+	return ColdWrite
 }
 
-func (b *dbBuffer) NeedsDrain() bool {
-	// Avoid capturing any variables with callback
-	return b.computedForEachBucketAsc(computeBucketIdx, bucketNeedsDrain) > 0
+func (b *dbBuffer) IsEmpty() bool {
+	// A buffer can only be empty if there are no buckets in its map, since
+	// buckets are only created when a write for a new block start is done, and
+	// buckets are removed from the map when they are evicted from memory.
+	return len(b.bucketsMap) == 0
 }
 
-func bucketNeedsDrain(now time.Time, b *dbBuffer, idx int, start time.Time) int {
-	if b.buckets[idx].needsDrain(now, start) {
-		return 1
+func (b *dbBuffer) Stats() bufferStats {
+	return bufferStats{
+		wiredBlocks: len(b.bucketsMap),
 	}
-	return 0
 }
 
 func (b *dbBuffer) Tick() bufferTickResult {
```

Comment:
Tick probably need to evict the bootstrapped blocks too right?

Reply:
Yep, that's in the `nonEvictedBuckets` part of the code. Bootstrapped blocks and incoming writes both live in the same bucket (the writable bucket).